libs/core/libs/shared/src/document_repo.rs

```diff
@@ -1,15 +1,155 @@
 use crate::core_config::Config;
-use crate::crypto::*;
 use crate::file_metadata::DocumentHmac;
-use crate::{SharedErrorKind, SharedResult};
-use std::collections::HashSet;
+use crate::SharedResult;
+use crate::{crypto::*, SharedErrorKind};
+use serde::{Deserialize, Serialize};
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
 use std::convert::TryInto;
 use std::fs::{self, File, OpenOptions};
+use std::hash::Hash;
 use std::io::{ErrorKind, Read, Write};
 use std::path::Path;
 use tracing::*;
 use uuid::Uuid;
 
+#[derive(Debug, Clone, Serialize, Deserialize, Ord, PartialEq, PartialOrd, Eq, Hash)]
+pub enum DocEvent {
+    Read(Uuid, i64),
+    Write(Uuid, i64),
+}
+impl DocEvent {
+    pub fn timestamp(&self) -> i64 {
+        match *self {
+            DocEvent::Read(_, timestamp) => timestamp,
+            DocEvent::Write(_, timestamp) => timestamp,
+        }
+    }
+    pub fn id(&self) -> Uuid {
+        match *self {
+            DocEvent::Read(id, _) => id,
+            DocEvent::Write(id, _) => id,
+        }
+    }
+}
+
+#[derive(Default, Copy, Clone, PartialEq)]
+pub struct StatisticValue {
+    pub raw: i64,
+    pub normalized: f64,
+}
+impl StatisticValue {
+    fn from_raw(raw: i64) -> StatisticValue {
+        StatisticValue { raw, normalized: f64::default() }
+    }
+}
+impl Ord for StatisticValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.raw).cmp(&other.raw)
+    }
+}
+
+impl PartialOrd for StatisticValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for StatisticValue {}
+
+#[derive(Clone, Copy)]
+pub struct StatisticValueRange {
+    pub max: StatisticValue,
+    pub min: StatisticValue,
+}
+impl StatisticValue {
+    pub fn normalize(&mut self, range: StatisticValueRange) {
+        self.normalized = (self.raw - range.min.raw) as f64 / (range.max.raw - range.min.raw) as f64
+    }
+}
+/// DocActivityMetrics stores key activity features for every doc which are used to recommend relevant documents
+/// latest_read_timestamp: the latests epoch timestamp that the user read a document
+/// latest_write_timestamp: the latests epoch timestamp that the user wrote a document
+/// the total number of times that a user read a document
+/// the total number of times that a user wrote a document   
+#[derive(Default, Copy, Clone)]
+pub struct DocActivityMetrics {
+    pub id: Uuid,
+    pub latest_read_timestamp: StatisticValue,
+    pub latest_write_timestamp: StatisticValue,
+    pub read_count: StatisticValue,
+    pub write_count: StatisticValue,
+}
+impl DocActivityMetrics {
+    pub fn rank(&self) -> i64 {
+        (self.latest_read_timestamp.normalized + self.latest_write_timestamp.normalized) as i64 * 70
```

Comment:
@Parth, What do you think about exposing the weights (70%, 30%) in the core endpoint as parameters to give flexibility to the caller? That way, this endpoint can serve as both suggested documents and most recent (if you pass the timestamp weight to 0), most accessed(if you pass the write/read count as 0 )

Reply:
I think we should expose these weights when clients want to customize them / expose them to users, for now it's fine